examples/parallel_bert.py

```diff
@@ -0,0 +1,192 @@
+                trunc = s[-maxlen:]
+            elif truncating == "post":
+                trunc = s[:maxlen]
+            else:
+                raise ValueError(f'Truncating type "{truncating}" not understood')
+
+            # check `trunc` has expected shape
+            trunc = np.asarray(trunc, dtype=dtype)
+            if trunc.shape[1:] != sample_shape:
+                raise ValueError(
+                    f"Shape of sample {trunc.shape[1:]} of sequence at "
+                    f"position {idx} is different from expected shape "
+                    f"{sample_shape}"
+                )
+
+            if padding == "post":
+                x[idx, : len(trunc)] = trunc
+            elif padding == "pre":
+                x[idx, -len(trunc) :] = trunc
+            else:
+                raise ValueError(f'Padding type "{padding}" not understood')
+        return x
+
+
+    input_ids = pad_sequences(input_ids, maxlen=MAX_LEN, dtype="long",
+                            value=0, truncating="post", padding="post")
+
+    train_inputs, validation_inputs, train_labels, validation_labels = train_test_split(input_ids, labels, random_state=13, test_size=0.1)
+
+    dl_builder = DataLoaderBuilder((FromArray(train_inputs, train_labels), \
+                                FromArray(validation_inputs, validation_labels)))
+
+    dl_tr, dl_val, _ = dl_builder.build(({"batch_size": args.batch_size}, {"batch_size": args.batch_size}))
+
+    if args.big_model:
+        model = BertForSequenceClassification.from_pretrained(
+            "bert-large-uncased",
+            num_labels = 2,
+            output_attentions = True,
+            output_hidden_states = False
+            )
+    else:
+        model = BertForSequenceClassification.from_pretrained(
+            "bert-base-uncased",
+            num_labels = 2,
+            output_attentions = True,
+            output_hidden_states = False
+            )
+
+    # Define the trainer
+
+    writer = GiottoSummaryWriter()
+    loss_function = nn.CrossEntropyLoss()
+    trainer = Trainer(model, (dl_tr, dl_val), loss_function, writer)
+    devices = list(range(torch.cuda.device_count()))
+    config_fsdp = {
+        "sharding_strategy": args.sharding.to_ss(),
+        "auto_wrap_policy": functools.partial(transformer_auto_wrap_policy, transformer_layer_cls={BertLayer,}),
+        "backward_prefetch": BackwardPrefetch.BACKWARD_PRE
+        }
+    parallel = Parallelism(args.parallel,
+                           devices,
+                           len(devices),
+                           config_fsdp=config_fsdp,
+                           pipeline_chunks=2)
+
+    # train the model
+
+    return trainer.train(Adam, args.n_epochs, parallel=parallel)
+
+
+if __name__ == '__main__':
+
+    parser = argparse.ArgumentParser(description='BERT Example')
+    gdeep.utility_examples.args.add_default_arguments(parser)
+    gdeep.utility_examples.args.add_big_model(parser)
+    parser.add_argument("--dl",
```

Comment:
This abbreviation probably stands for `download` but could be confused with `dataloader` since it's used like that at other places.

Reply:
It will be improved in the next push.